File: backend/src/integrations/ticketing/jira_connector.js
/**
 * Jira Ticketing Connector — Phase 14.1
 *
 * Implements connector interface for Atlassian Jira Cloud and Jira Data Center/Server.
 * Includes all 8 mandatory fields:
 * - asset ID
 * - finding ID
 * - severity
 * - owner
 * - evidence link
 * - remediation
 * - CBOM reference
 * - risk score
 */

const { BaseTicketingConnector } = require("./base_connector");
const { TicketResponse } = require("../../domain/contracts");

const DEFAULT_PRIORITY_MAP = {
  CRITICAL: "Highest",
  HIGH: "High",
  MEDIUM: "Medium",
  LOW: "Low",
  INFO: "Lowest",
  INFORMATIONAL: "Lowest",
};

class JiraConnector extends BaseTicketingConnector {
  constructor(options = {}) {
    super({
      name: options.name || "jira-default",
      connectorType: "jira",
      config: options.config || {},
      fetchFn: options.fetchFn,
    });
  }

  validateConfig(config) {
    super.validateConfig(config);
    if (!config.host) throw new Error("Jira connector requires 'host' URL");
    if (!config.projectKey) throw new Error("Jira connector requires 'projectKey'");
    if (!config.apiToken) throw new Error("Jira connector requires 'apiToken'");
    if (!config.username) throw new Error("Jira connector requires 'username' (email/service account)");
  }

  get normalizedHost() {
    return this.config.host.replace(/\/+$/, "");
  }

  getAuthHeader() {
    const creds = Buffer.from(`${this.config.username}:${this.config.apiToken}`).toString("base64");
    return `Basic ${creds}`;
  }

  mapPriority(severity) {
    const sev = String(severity || "MEDIUM").toUpperCase();
    const customMap = this.config.priorityMapping || {};
    return customMap[sev] || DEFAULT_PRIORITY_MAP[sev] || "Medium";
  }

  /**
   * Pure formatter converting TicketRequest into Jira REST API v2/v3 issue payload.
   */
  formatPayload(ticketRequest) {
    const priority = this.mapPriority(ticketRequest.severity);
    const issueType = this.config.issueType || "Bug";

    // Build Jira wiki-markup description containing all 8 mandatory fields
    const descriptionLines = [
      "h2. ECDAT Cryptographic Finding Details",
      "",
      "||Field||Value||",
      `|*Asset ID*|${ticketRequest.assetId}|`,
      `|*Finding ID*|${ticketRequest.findingId}|`,
      `|*Severity*|${ticketRequest.severity}|`,
      `|*Owner*|${ticketRequest.owner}|`,
      `|*Evidence Link*|[${ticketRequest.evidenceLink}|${ticketRequest.evidenceLink}]|`,
      `|*CBOM Reference*|${ticketRequest.cbomRef}|`,
      `|*Risk Score*|${ticketRequest.riskScore}|`,
      "",
      "h3. Remediation Guidance",
      ticketRequest.remediation,
      "",
      "----",
      "_Generated automatically by ECDAT (Enterprise Cryptographic Discovery & Agility Toolkit)_",
    ];

    const fields = {
      project: { key: this.config.projectKey },
      summary: ticketRequest.title || `[${ticketRequest.severity}] Cryptographic Finding: ${ticketRequest.findingId} on ${ticketRequest.assetId}`,
      description: descriptionLines.join("\n"),
      issuetype: { name: issueType },
      priority: { name: priority },
      labels: [
        "ecdat",
        "cryptography",
        "security",
        `sev-${ticketRequest.severity.toLowerCase()}`,
      ],
    };

    // Custom fields if configured
    if (this.config.customFields && typeof this.config.customFields === "object") {
      for (const [k, v] of Object.entries(this.config.customFields)) {
        fields[k] = v;
      }
    }

    return { fields };
  }

  async sendCreateRequest(ticketRequest, payload) {
    const endpoint = `${this.normalizedHost}/rest/api/2/issue`;
    const headers = {
      "Content-Type": "application/json",
      Authorization: this.getAuthHeader(),
      Accept: "application/json",
    };

    const res = await this.fetchFn(endpoint, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
    });

    const data = await res.json().catch(() => ({}));
    if (!res.ok) {
      const errMsg = data.errorMessages ? data.errorMessages.join("; ") : JSON.stringify(data.errors || data);
      throw new Error(`Jira ticket creation failed (HTTP ${res.status}): ${errMsg}`);
    }

    const ticketKey = data.key || data.id;
    const ticketUrl = `${this.normalizedHost}/browse/${ticketKey}`;

    return new TicketResponse({
      success: true,
      ticketId: ticketKey,
      ticketUrl,
      connectorType: "jira",
      status: "OPEN",
      rawResponse: data,
    });
  }

  async testConnection() {
    const endpoint = `${this.normalizedHost}/rest/api/2/myself`;
    const res = await this.fetchFn(endpoint, {
      method: "GET",
      headers: {
        Authorization: this.getAuthHeader(),
        Accept: "application/json",
      },
    });

    if (!res.ok) {
      throw new Error(`Jira connection test failed with status ${res.status}`);
    }

    const data = await res.json().catch(() => ({}));
    return {
      ok: true,
      message: `Successfully connected to Jira host '${this.normalizedHost}' as '${data.displayName || data.name || this.config.username}'`,
    };
  }
}

module.exports = {
  JiraConnector,
  DEFAULT_PRIORITY_MAP,
};

File: scanners/integrations/jira_connector.py
"""
Jira Ticketing Connector (Python) — Phase 14.1

Implements connector interface for Atlassian Jira Cloud & Server.
Includes all 8 mandatory fields:
- asset ID
- finding ID
- severity
- owner
- evidence link
- remediation
- CBOM reference
- risk score
"""

import base64
import json
from typing import Any, Callable, Dict, Optional
import urllib.request
import urllib.error

from scanners.integrations.base_connector import BaseTicketingConnector
from scanners.integrations.ticket_request import TicketRequest, TicketResponse

DEFAULT_PRIORITY_MAP = {
    "CRITICAL": "Highest",
    "HIGH": "High",
    "MEDIUM": "Medium",
    "LOW": "Low",
    "INFO": "Lowest",
    "INFORMATIONAL": "Lowest",
}


class JiraConnector(BaseTicketingConnector):
    def __init__(
        self,
        name: str = "jira-default",
        config: Optional[Dict[str, Any]] = None,
        http_client: Optional[Callable[..., Any]] = None,
    ):
        super().__init__(
            name=name,
            connector_type="jira",
            config=config,
            http_client=http_client,
        )

    def validate_config(self, config: Dict[str, Any]) -> None:
        super().validate_config(config)
        if not config.get("host"):
            raise ValueError("Jira connector requires 'host' URL")
        if not config.get("project_key") and not config.get("projectKey"):
            raise ValueError("Jira connector requires 'project_key'")
        if not config.get("api_token") and not config.get("apiToken"):
            raise ValueError("Jira connector requires 'api_token'")
        if not config.get("username"):
            raise ValueError("Jira connector requires 'username'")

    @property
    def normalized_host(self) -> str:
        return str(self.config.get("host", "")).rstrip("/")

    @property
    def project_key(self) -> str:
        return self.config.get("project_key") or self.config.get("projectKey", "")

    @property
    def api_token(self) -> str:
        return self.config.get("api_token") or self.config.get("apiToken", "")

    @property
    def username(self) -> str:
        return self.config.get("username", "")

    def get_auth_header(self) -> str:
        raw = f"{self.username}:{self.api_token}".encode("utf-8")
        return f"Basic {base64.b64encode(raw).decode('utf-8')}"

    def map_priority(self, severity: str) -> str:
        sev = str(severity or "MEDIUM").upper()
        custom = self.config.get("priority_mapping") or self.config.get("priorityMapping") or {}
        return custom.get(sev, DEFAULT_PRIORITY_MAP.get(sev, "Medium"))

    def format_payload(self, ticket_request: TicketRequest) -> Dict[str, Any]:
        priority = self.map_priority(ticket_request.severity)
        issue_type = self.config.get("issue_type") or self.config.get("issueType") or "Bug"

        description_lines = [
            "h2. ECDAT Cryptographic Finding Details",
            "",
            "||Field||Value||",
            f"|*Asset ID*|{ticket_request.asset_id}|",
            f"|*Finding ID*|{ticket_request.finding_id}|",
            f"|*Severity*|{ticket_request.severity}|",
            f"|*Owner*|{ticket_request.owner}|",
            f"|*Evidence Link*|[{ticket_request.evidence_link}|{ticket_request.evidence_link}]|",
            f"|*CBOM Reference*|{ticket_request.cbom_ref}|",
            f"|*Risk Score*|{ticket_request.risk_score}|",
            "",
            "h3. Remediation Guidance",
            ticket_request.remediation,
            "",
            "----",
            "_Generated automatically by ECDAT (Enterprise Cryptographic Discovery & Agility Toolkit)_",
        ]

        fields = {
            "project": {"key": self.project_key},
            "summary": ticket_request.title
            or f"[{ticket_request.severity}] Finding: {ticket_request.finding_id} on {ticket_request.asset_id}",
            "description": "\n".join(description_lines),
            "issuetype": {"name": issue_type},
            "priority": {"name": priority},
            "labels": [
                "ecdat",
                "cryptography",
                "security",
                f"sev-{ticket_request.severity.lower()}",
            ],
        }

        custom_fields = self.config.get("custom_fields") or self.config.get("customFields")
        if isinstance(custom_fields, dict):
            fields.update(custom_fields)

        return {"fields": fields}

    def send_create_request(self, ticket_request: TicketRequest, payload: Dict[str, Any]) -> TicketResponse:
        endpoint = f"{self.normalized_host}/rest/api/2/issue"

        if self.http_client:
            data = self.http_client(
                endpoint, method="POST", json_payload=payload, headers={"Authorization": self.get_auth_header()}
            )
        else:
            req = urllib.request.Request(
                endpoint,
                data=json.dumps(payload).encode("utf-8"),
                headers={
                    "Content-Type": "application/json",
                    "Authorization": self.get_auth_header(),
                    "Accept": "application/json",
                },
                method="POST",
            )
            try:
                # nosemgrep: python.lang.security.audit.dynamic-urllib-use-detected.dynamic-urllib-use-detected
                with urllib.request.urlopen(req) as resp:  # nosec B310
                    data = json.loads(resp.read().decode("utf-8"))
            except urllib.error.HTTPError as e:
                err_text = e.read().decode("utf-8")
                raise RuntimeError(f"Jira API error ({e.code}): {err_text}")

        ticket_key = data.get("key") or data.get("id") or "UNKNOWN"
        ticket_url = f"{self.normalized_host}/browse/{ticket_key}"

        return TicketResponse(
            success=True,
            ticket_id=str(ticket_key),
            ticket_url=ticket_url,
            connector_type="jira",
            status="OPEN",
            raw_response=data,
        )
